ExpandableBlock props for subcomponents (#2460)
iTwin/iTwinUI: apps/website/src/content/docs/expandableblock.mdx

## Usage

> ### Legacy API

Author: As suggested in https://github.com/iTwin/iTwinUI/pull/2460#discussion_r2017688450, I have added to the usage section to describe the legacy and composition API. Please let me know if this should be adjusted or changed.

Reviewer: In my [suggestion](https://github.com/iTwin/iTwinUI/pull/2460#discussion_r2025254378), I renamed it to "Props API" because I don't know if we should consider this "legacy" right now.

You can use `ExpandableBlock` directly, without any subcomponents.

> - Directly using `ExpandableBlock` along with `title`, `caption`, `endIcon` and children.

Author: Changes have been made.

> - Although this pattern can suffice for simple cases, we strongly recommend using composition API for more flexibility.

Author: Just to clarify, changes have been made so that the second bullet point is now a `**Note**`. If I misinterpreted the feedback, please let me know.

> - Although this pattern can suffice for simple cases, we strongly recommend using composition API for more flexibility.

Reviewer: In my [suggestion](https://github.com/iTwin/iTwinUI/pull/2460#discussion_r2025254378), I moved the Note to the bottom and removed the word "strongly" for the same reason as https://github.com/iTwin/iTwinUI/pull/2460#discussion_r2025266850. Not sure if we have a strong reason to discourage the props API. This component is different from `Tabs`, where we do indeed strongly recommend the composition API.

## Usage

> ### Legacy API
> 
> **Note**: Although this pattern can suffice for simple cases, we strongly recommend using composition API for more flexibility.
> 
> You can use `ExpandableBlock` directly, without any subcomponents.
> 
> - Directly using `ExpandableBlock` along with `title`, `caption`, `endIcon` and `children`.
> 
> ```tsx
> <ExpandableBlock title='Expandable Block' caption='With caption!'>
>   Content in block!
> </ExpandableBlock>
> ```

Reviewer: ````suggestion
### Props API

The `<ExpandableBlock>` component can be used directly, without any subcomponents.

It accepts `title`, `caption`, `endIcon` and `children` props.

```tsx
<ExpandableBlock title='Expandable Block' caption='With caption!'>
  Content in block!
</ExpandableBlock>
```

**Note**: Although this pattern can suffice for simple cases, we strongly recommend using composition API for more flexibility.
````

Author: Changes made.

### Composition API

> Alternatively, you can build expandable blocks with sub-components that are fully customizable.

Reviewer: ```suggestion
Alternatively, you can use subcomponents that are fully customizable. This pattern is more verbose but allows full customization over all the parts.
```

Author: Changes made.

> - Using subcomponents (`ExpandableBlock.Wrapper`, `ExpandableBlock.Trigger`, `ExpandableBlock.Content`, etc). This pattern is more verbose but allows full customization over all the parts.
>   - `ExpandableBlock.Wrapper` is a wrapper around the subcomponents.
>   - The `ExpandableBlock.Trigger` is the `Header` container that contains `ExpandIcon`, `LabelArea` and `EndIcon` subcomponents.
>   - The `ExpandableBlock.Content` is the content shown when `ExpandableBlock.Trigger` is triggered.
> 
> The following code is equivalent to the [legacy example](#legacy-api) above:

Reviewer: ```suggestion
 - `<ExpandableBlock.Wrapper>` is a wrapper around the subcomponents.
 - `<ExpandableBlock.Trigger>` is the "trigger" element that's always shown.
    - `<ExpandableBlock.ExpandIcon>`, `<ExpandableBlock.LabelArea>` and `<ExpandableBlock.EndIcon>` can be used inside `<ExpandableBlock.Trigger>`.
    - `<ExpandableBlock.Title>` and `<ExpandableBlock.Caption>` can be used inside `<ExpandableBlock.LabelArea>`.
    - Alternatively, `label`, `caption`, `expandIcon`, and `endIcon` props can be passed to `<ExpandableBlock.Trigger>`.
 - `<ExpandableBlock.Content>` is the content that gets revealed `<ExpandableBlock.Trigger>` is clicked.

The following code is equivalent to the [legacy example](#legacy-api) above:
```

Author: Changes made with minor adjustments (adding missing "when" in last bullet, changing legacy to props in last line of suggestion).

The expandable block can be used in two ways:

- Directly using `ExpandableBlock` along with `title`, `caption`, `endIcon` and children. This pattern can suffice for simple cases.

>   - `ExpandableBlock` has customizable `Title`, `Caption`, `Content` and `EndIcon` subcomponents.

Reviewer: This is incorrect and misleading.

```suggestion
```

Author: The bullet has been removed. If this is the case, should the corresponding description be handled as well in line 341 in `ExpandableBlock.tsx`?

Reviewer: That particular line, while not inaccurate, could be improved to clarify the usage.

The key thing to understand is that `ExpandableBlock` can be used either standalone without any subcomponents (`<ExpandableBlock>`) or with only subcomponents (`<ExpandableBlock.Wrapper>` and children). These are basically two different components, even though they are handled by a single combined export.

I believe we use the term "Props API" (or "Legacy API") and "Composition API" for these two different kinds of usage. See [Alert](https://itwinui.bentley.com/docs/alert/#usage), [Tabs](https://itwinui.bentley.com/docs/tabs/#usage) and [Tile](https://itwinui.bentley.com/docs/tile/).

## Props

### ExpandableBlock

`ExpandableBlock` with customizable `Title`, `Caption`, `Content` and `EndIcon` subcomponents.

<PropsTable path='@itwin/itwinui-react/esm/core/ExpandableBlock/ExpandableBlock.d.ts' />

> ## Subcomponents

Author: Changes made.

### ExpandableBlock.Wrapper

> `ExpandableBlock.Wrapper` should contain the `ExpandableBlock` and the component over which the block will overlay.

Author: Changes made.

> 

Author: Yeah, thanks for pointing that out. I can change it to be clearer as you described.

> 

Author: Changes made.

> <PropsTable

Author: Added the link to the related discussion for this PR in the description. The comments mainly were about these particular subcomponents. However, if it would make sense to do this for all subcomponents, the relevant changes can be made.

  componentName='ExpandableBlockWrapper'
  path='@itwin/itwinui-react/esm/core/ExpandableBlock/ExpandableBlock.d.ts'
/>

### ExpandableBlock.Trigger

The `ExpandableBlock.Trigger` is the `Header` container that contains `ExpandIcon`, `LabelArea` and `EndIcon` subcomponents.

<PropsTable
  componentName='ExpandableBlockTrigger'
  path='@itwin/itwinui-react/esm/core/ExpandableBlock/ExpandableBlock.d.ts'
/>

### ExpandableBlock.Content

> The `ExpandableBlock.Content` subcomponent is used for the content shown in the block's body when fully expanded.

Author: Changes made.
